### PR TITLE
Improve build scan metadata

### PR DESCRIPTION
**Pull request by mark-vieira**
This PR does a couple of things:

First, we move the logic for some of our custom build scan into the build itself vs having this live in our Jenkins configuration. Having the conditional tagging logic in our Jenkins config is awkward for a few reasons.

1. Any actual conditional logic has to be done in bash, rather than in Groovy.
2. There's a good deal of duplication as this gets copied to all our builder macros.
3. For any complex custom values we have to workaround Jenkins odd handling of string escaping which doesn't often align with normal shell escaping rules.
4. We can't easily add custom data computed from actual build configuration.

Second, we've added some addition bits of information, mainly to fill in context gaps with the goal of making it less necessary to have to go back to the original Jenkins job page to get that information.

1. The actual git commit, as well as link to the commit info on GitHub.
2. The git branch.
3. For CI builds, include the list of changes included in the job run.

### CODE REVIEW

File: gradle/build-scan.gradle [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+buildScan {
+    // Accept Gradle ToS when project property org.elasticsearch.acceptScanTOS=true
+    if (Boolean.valueOf(project.findProperty('org.elasticsearch.acceptScanTOS') ?: "false")) {
+        termsOfServiceUrl = 'https://gradle.com/terms-of-service'
+        termsOfServiceAgree = 'yes'
+    }
+
+    // Jenkins-specific build scan metadata
+    if (System.getenv('JENKINS_URL')) {
```
Comment (atorok) [inline gradle/build-scan.gradle]: We could check if this contains `elastic.co` to make sure it's our CI and accept the terms of service if so so no additional property needs to be passed in Jenkins.

**Review by alpar-t — CHANGES_REQUESTED** [review]
Looks good, but left some suggestions. 

While at  it it would be great to have information about the CI worker this is what I most often have to go back to jenkins for. 
Jenkins sets `NODE_LABELS` we could prefix them and add them as tags here. 

Instead of an included file, I'm wondering if it would be better to start a base plugin that would apply configuration to the root plugin and have this live in buildSrc along with everything else.  We're starting to have more and more configuration that applies to the root project only.

File: gradle/build-scan.gradle (hunk 1 above)
Comment (mark-vieira) [reply gradle/build-scan.gradle]: Eventually we'll be pushing scans to Gradle Enterprise which doesn't require accepting the ToS so I think that might be unnecessary.

**Comment by mark-vieira** [comment]
> While at it it would be great to have information about the CI worker this is what I most often have to go back to jenkins for.

Indeed. Is it the labels specifically that you find yourself looking for? Is the information in the [infrastructure](https://scans.gradle.com/s/hlct5s3dlxiyk/infrastructure) section of the build scan insufficient?

**Comment by mark-vieira** [comment]
> Instead of an included file, I'm wondering if it would be better to start a base plugin that would apply configuration to the root plugin and have this live in buildSrc along with everything else. We're starting to have more and more configuration that applies to the root project only.

Yeah, I definitely didn't want to expand the root `build.gradle` any further. We should certainly work towards trimming that down and moving that logic into plugins of some sort, whether script plugins or binary plugins in `buildSrc` is almost a matter of preference.

File: gradle/build-scan.gradle (hunk 1 above)
Comment (mark-vieira) [reply gradle/build-scan.gradle]: I've gone ahead and implemented this. I agree, it'd be nice to ditch the need to pass the ToS param in Jenkins for the time being.

**Comment by mark-vieira** [comment]
> Jenkins sets NODE_LABELS we could prefix them and add them as tags here.

I've added these as custom values rather than tags to try and minimize a bit of noise as some of these are long (like the work name) and there can be lots of them. We can still search and filter builds in Gradle Enterprise with this method though. So we could say "show me all build that ran on a worker with the `metal` label".

**Comment by mark-vieira** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by mark-vieira** [comment]
@atorok here's an example of the node labels: https://scans.gradle.com/s/hnaify7kg33vg/custom-values#L2-L5

File: gradle/build-scan.gradle [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import nebula.plugin.info.scm.ScmInfoExtension
+
+buildScan {
+    def jenkinsUrl = System.getenv('JENKINS_URL') ? new URL(System.getenv('JENKINS_URL')) : null
+
+    // Accept Gradle ToS when project property org.elasticsearch.acceptScanTOS=true or this is an Elastic CI build
+    if (jenkinsUrl?.host?.endsWith('elastic.co') || Boolean.valueOf(project.findProperty('org.elasticsearch.acceptScanTOS') ?: "false")) {
```
Comment (atorok) [inline gradle/build-scan.gradle]: we could just ditch the property, I don't see a reason to opt out of CI

**Review by alpar-t — APPROVED** [review]

File: gradle/build-scan.gradle (hunk 2 above)
Comment (mark-vieira) [reply gradle/build-scan.gradle]: I use this property to opt-in locally.

**Pull request merged**